cluster-autoscaler/proposals/gpu_optimization.md

```diff
@@ -0,0 +1,88 @@
+# Cluster Autoscaling Optimization for GPU clusters
+##### Author: Jeffwan
+
+## Introduction
+Cluster Autoscaler makes it extremely easy to scale kubernetes cluster in response to pod status. At the same time, having it automate Kubernetes cluster to work efficiently.
+
+However, there's still challenges in specific scenarios for accelerator computing clusters. Accelerator computing instances are quite different from normal instance. They are powerful, scalable instances that provide GPU-based parallel computing capabilities. 
+
+Accelerator computings nodes are expensive. (ec2 p3.16xlarge on demand instance cost $24.48/hr). Efficiently scale the cluster will save lots of money. GPU nodes are well suited for tasks that have heavy computation needs like machine learning and HPC applications. They can not be interrupted in some cases and that brings us more challenges in scale down. 
+
+Here's the problems I find in current upstream cluster autoscaler. 
+
+### Problems in cluster autoscaler for GPU node
+
+#### Accelerator Label 
+GKE uses gpu label `cloud.google.com/gke-accelerator` to inspect if a node is an accelerator node, other cloud providers don't define their label yet. In order for CA to behavior normally in GPU cluster, it's essential to have this gpu label.
+
+#### Scale too much without GPU label support
+Pending pods will trigger scale up decision to bring up number of nodes in one node group. When new node is added, CA knows it is `upcoming` and it knows some pods will go on that node once it boots up. So it won't trigger another scale-up for those pods. Device plugin triggers after the node becomes ready and the extra resource required by pod will show up in node spec even later. CA will see that it was wrong and the pods for which it triggered scale-up can't actually go to a node it added for them (because it doesn't have the special resource yet and CA doesn't know it will be added later). So CA will go and create more nodes.
+
+The problem is GPU node becomes ready first but at that time, there's no allocable GPU resources. The root reason could be  
+
+* Device Plugin is not ready.
+* Device Plugin is ready but has not sent list of devices to kubelet. 
+* kubelet takes time to advertise resources to API server so that node spec is not up to date.
+
+Pod thhat requests GPU can not be scheduled during this period. In order to address this issue, we either need to work with Nvidia folks on device plugin to change behavior on upstream, or make CA mark nodes with GPU label and unallocated GPU resource as NonReady.
+
+
+#### GPU Resource is not given consideration in scaling down
+CA filter candidates with low node cpu and memory utilization rate in every loop. 
+
+* If utilization rate is high, even all GPUs are idle, scale-down will not be triggered.  
```

Comment:
1. Agree. This is what user normally do for training jobs. There's a case that I am not sure if that's common enough. For tensorflow distributed training, parameter server and workers are just deployed on one GPU cluster but PS pod utilize more on CPU and memory and worker utilize GPU. 

2. "If we're talking only about GPU workloads that could be spread across other nodes blocking scale-down," That's the inference case and workloads could mix together on some economic GPU nodes.

Reply:
I agree with safe-to-evict solution for training jobs.

GPU resource is not even considered in the utilization calculation here. So raising utilization threshold won't help this case. 
https://github.com/kubernetes/autoscaler/blob/14ed6ecb9281b89dc27f6cb632d2e07b22092f31/cluster-autoscaler/simulator/cluster.go#L155-L164

I think this part we can do some improvement.

Reply:
To be clear: this comment is to a section of the doc concerned with idle GPUs + other pods using up CPU/memory above threshold. I'm fairly certain raising threshold will help with that.

I'm not sure what kind of benefit is to be gained from adding GPU utilization into the mix, though?

Reply:
I don't think we should look into CPU / memory utilization on nodes with GPUs at all. There are valid use-cases where you can have pods taking significant chunk of CPU on a node with GPU, that can't be addressed by separating workloads. One example is https://github.com/kubernetes/autoscaler/issues/1367 where it was the driver daemonset using >50% of CPU, the other is things like monitoring infrastructure taking a significant chunk of CPU/memory. Bottom line is: you don't want to keep GPU node around just because it's CPU is highly utilized. If there is a pod that can't be restarted drain will catch it, but CPU utilization itself is not enough reason to keep it around.

You raise a good point though - we could just completely skip utilization check for GPU nodes. That being said it may be useful for the same reason it's useful for CPU: you may want to allow restarting a job if the cost saving is very significant (low utilization), but not if the saving is minimal (high utilization, but the pods can be drained). At the end of the day you can always ignore utilization by setting threshold to 100%. So I think the most correct thing would be to:
1. Add a GPU utilization check for nodes with GPU. I think it should get a separate threshold as well. 
2. Ignore CPU / memory checks on nodes with GPU.

Reply:
Totally Agree.

Reply:
In most cases, even if CPU utilization is low for a pod on a node-bearing GPU, that it's using the GPU at all says that "moving" it from that node is not likely any better; at best, it just induces needless churn.

Most of our machines have one or two GPUs, and I've seen the autoscaler declare that a pod using the lone GPU on a node is underutilizing the resources, as if it could "move" the pod elsewhere and have it use resources more efficiently. However, all the _other_ nodes also have only one GPU. A replacement pod can't land anywhere else that a similar pod is present using the only GPU. We wind up losing our nodes and needing to replace them with fresh nodes immediately afterward.

Reply:
@seh sounds like a good use-case for `safe-to-evict: false`?

Reply:
That works from the perspective of the maintainers of this project, but it's not fair to _users_—users who may not even know that the cluster is using an autoscaler, let alone this one.

They write a _Job_ or _Deployment_ with a pod spec that says, "I need one GPU." That is honest and true. The admission control plugin augments the pod with the corresponding toleration, the autoscaler trusts that request and summons a suitable machine, the scheduler honors that request and places the pod, and the pod uses the GPU as promised. Then the autoscaler comes back around and says, "I don't care what you say, you're wasting resources. I'm kicking you off."

Why should the pod author need to learn about how to convince this particular autoscaler that not only would he like a node with a GPU, he'd also like to _stay_ on that node with a GPU while he's using it. There's no disruption going on here, and there's no waste. The machine has one GPU, and the pod is using one GPU. Why demand use of an annotation to say, "Really, I meant what I said?"

Reply:
> Why demand use of an annotation to say, "Really, I meant what I said?"

OK, how should the autoscaler determine which pods to move and which not to move? Like @Jeffwan mentioned earlier, in some cases it makes sense to reschedule GPU pods. Both kind of jobs (evictable and non-evictable) may run the same cluster, especially if it's shared across teams. For you, just using GPUs may mean you don't want the pod evicted, ever - but this may simply not work for other users.

Whether rescheduling Job pods is optimal as a default is another question altogether. There is definitely some negative feedback, but then again, users who are satisfied with something are much less likely to. It's possible we made a mistake here, but changing the default at this point would probably be very confusing and disruptive, especially for the users who don't care about cluster administration and/or don't control cluster upgrades. 

As for putting extra toll on the pod owner, if this becomes a significant problem, it can probably be addressed similarly to adding a toleration you described - by admission controller or other automation, which can add annotations or set up PDBs (maxUnavailable: 0 should have the same effect).

Reply:
@seh Your argument applies to any workload, regardless if it uses GPU or not. It's autoscalers job to kick pods around to limit resource usage. An application may not agree to be restarted in which case it must specify it somehow.

> Why should the pod author need to learn about how to convince this particular autoscaler 

PDB is a generic Kubernetes way of telling all automation (including CA) not to touch a given pod. The annotation exists because many people requested a way to specifically prevent CA from touching the pod, but allow other automation to restart it.

> The machine has one GPU, and the pod is using one GPU

Which means GPU utilization for node is 100%. Once we add GPU utilization check as described in comments above CA will not touch that node. You'll also be able to set threshold to 0, in which case no node will be removed if any of its GPUs are used.

Reply:
> OK, how should the autoscaler determine which pods to move and which not to move? [...] For you, just using GPUs may mean you don't want the pod evicted, ever - but this may simply not work for other users.

The autoscaler normally doesn't "move" pods unless it thinks they're not utilizing a machine's resources adequately. That situation does not—or should not—apply here.

In my example, we have a machine with one GPU, a pod running there that requests _and is using_ one GPU, and no other nodes advertising available GPUs. Why would an autoscaler consider this pod as a candidate to move? By analogy, if a machine has one CPU and a pod there is "using" (really, requested) the whole CPU, the autoscaler wouldn't consider moving that pod. Why do we trust use of a CPU but not use of a GPU?

> it can probably be addressed similarly to adding a toleration you described - by admission controller or other automation, which can add annotations or set up PDBs (maxUnavailable: 0 should have the same effect).

The toleration I mentioned is a convenience provided by [the "ExtendedResourceToleration" admission control plugin](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#extendedresourcetoleration); we used to write them manually.

Reply:
> Which means GPU utilization for node is 100%. Once we add GPU utilization check as described in comments above CA will not touch that node.

Well, that's great news! That's what I'm asking for here. If we're going to count GPU utilization and not move pods that are utilizing the machine's resources adequately, then I have no further complaint.